GH-39010: [Python] Introduce `maps_as_pydicts` parameter for `to_pylist`, `to_pydict`, `as_py`

jonded94 opened the pull request:
### Rationale for this change

Currently, unfortunately `MapScalar`/`Array` types are not deserialized into proper Python `dict`s, which is unfortunate since this breaks "roundtrips" from Python -> Arrow -> Python:

```
import pyarrow as pa

schema = pa.schema([pa.field('x', pa.map_(pa.string(), pa.int64()))])
data = [{'x': {'a': 1}}]
pa.RecordBatch.from_pylist(data, schema=schema).to_pylist()
# [{'x': [('a', 1)]}]
```

This is especially bad when storing TiBs of deeply nested data (think of lists in structs in maps...) that were created from Python and serialized into Arrow/Parquet, since they can't be read in again with native `pyarrow` methods without doing extremely ugly and computationally costly workarounds.

### What changes are included in this PR?

A new parameter `maps_as_pydicts` is introduced to `to_pylist`, `to_pydict`, `as_py` which will allow proper roundtrips:

```
import pyarrow as pa

schema = pa.schema([pa.field('x', pa.map_(pa.string(), pa.int64()))])
data = [{'x': {'a': 1}}]
pa.RecordBatch.from_pylist(data, schema=schema).to_pylist(maps_as_pydicts="strict")
# [{'x': {'a': 1}}]
```

### Are these changes tested?

Yes. There are tests for `to_pylist` and `to_pydict` included for `pyarrow.Table`, whilst low-level `MapScalar` and especially a nesting with `ListScalar` and `StructScalar` is tested.

Also, duplicate keys now should throw an error, which is also tested for.

### Are there any user-facing changes?

No callsites should be broken, simply a new keyword-only optional parameter is added.
* GitHub Issue: #39010

pitrou commented:
While this is not a bad idea in itself, it seems like the roundtripping concern could be solved more efficiently by making `from_pylist` accept a list of tuples for map fields.

pitrou commented:
Also:
> they can't be read in again with native pyarrow methods without doing extremely ugly and computationally costly workarounds

Please note that `from_pylist` and `to_pylist` are quite costly in themselves. Usually you want to avoid these kinds of roundtrips to/from Python objects if you are concerned with performance.

jonded94 commented:
> While this is not a bad idea in itself, it seems like the roundtripping concern could be solved more efficiently by making from_pylist accept a list of tuples for map fields.

Let me clarify what this is about. Map fields are already createable with `from_pylist` by using list of tuples, as I show in the tests I added. Even the code in my initial message can show this. Fundamentally, it's about adding opt-in behaviour to `to_pylist` to arrive at a functionality one would expect from a Python perspective:

```
data = [{'x': {'a': 1}}]
pa.RecordBatch.from_pylist(data, schema=schema).to_pylist()
^---------------------------------------------^
  this works fine, data will properly encoded in the Arrow way of encoding Maps
                                                ^---------^
                                                this will give lists of tuples instead of dicts 
```

You can use `data = [{'x': [('a', 1)]}]` here too, this will yield the same `RecordBatch`. This then of course *technically* would qualify as a proper "roundtrip", but this is not what this issue is about, it's about deserializing Map Arrow types as the ~"expected" Python equivalent, at least as an opt-in method such as `pandas` already supports for some couple of years now (shown in the [linked Github issue](https://github.com/apache/arrow/issues/39010)).

> Please note that from_pylist and to_pylist are quite costly in themselves. 

Yes, but this is part of a very large distributed machine learning setup, where relatively intricate filters applied on deeply nested list/struct/map columns. The compute of the actual machine learning outclasses the compute one has to do to deserialize Python objects by many orders of magnitude.

For pure data queries, we would not use bare Python objects of course.

pitrou commented:
> it's about deserializing Map Arrow types as the ~"expected" Python equivalent, at least as an opt-in method such as `pandas` already supports for some couple of years now (shown in the [linked Github issue](https://github.com/apache/arrow/issues/39010)).

I see, thanks. Then, do we want to reuse the same parameter signature as in the Pandas-related PR? I.e., allow either `None`, `"lossy"` and `"strict"`, rather than a boolean.

jonded94 commented:
> allow either None, "lossy" and "strict", rather than a boolean.

Sure, I actually also stumbled across that when I revisited that original Github issue. Before I do that, I'd like to ask whether you're generally fine with adding this new parameter to *every* `as_py` method? This has to be done because the  `to_pylist` method calls `as_py` on its member arrays (which can be all possible types), and therefore all array/scalar types have to support this parameter. I did not see any other way to easily implement this. I'm willing to do quick progress here, so if you come up with another idea, let me know.

pitrou commented:
> Sure, I actually also stumbled across that when I revisited that original Github issue. Before I do that, I'd like to ask whether you're generally fine with adding this new parameter to _every_ `as_py` method?

That sounds ok to me. Ideally, `to_pylist` wouldn't call `as_py` in a loop (which is going to be quite slow), but that would be a major refactor.

pitrou commented:
By the way, we probably want to make the new parameter keyword-only?

jonded94 commented:
I addressed the remarks :) There is some weird error in the ["Docs" job](https://github.com/jonded94/arrow/actions/runs/13248562449/job/36980840032), I don't know what this is about.

jonded94 commented:
I rebased the branch, now the CI tests seem fine again, I think? 

Could we get a approval/review of this? :)

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -1651,16 +1651,32 @@ cdef class Array(_PandasConvertible):
+            Arrow Map, as in [(key1, value1), (key2, value2), ...].
+
+            If 'lossy' or 'strict', convert Arrow Map arrays to native Python dicts.
+            This can change the ordering of (key, value) pairs, and will
+            deduplicate multiple keys, resulting in a possible loss of data.
```
I think the ordering comment is obsolete, as Python dicts are ordered nowadays. Unless the underlying implementation does something weird, ordering should therefore be preserved.

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -2286,12 +2302,28 @@ cdef class MonthDayNanoIntervalArray(Array):
+        Parameters
+        ----------
+        maps_as_pydicts : str, optional, default `None`
+            Valid values are `None`, 'lossy', or 'strict'.
```
This is the docstring for `MonthDayNanoIntervalArray`, so `maps_as_pydicts` would simply be ignored here. How about just stating that?

jonded94 replied:
Added a message that this is ignored for non-nested Scalars.

pitrou commented on python/pyarrow/table.pxi:
```diff
@@ -1349,10 +1349,26 @@ cdef class ChunkedArray(_PandasConvertible):
+            Arrow Map, as in [(key1, value1), (key2, value2), ...].
+
+            If 'lossy' or 'strict', convert Arrow Map arrays to native Python dicts.
+            This can change the ordering of (key, value) pairs, and will
```
Same comment re: ordering

pitrou commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -786,6 +786,21 @@ def test_map(pickle_module):
+    with pytest.raises(ValueError):
+        assert s.as_py(maps_as_pydicts="strict")
+
+    assert s.as_py(maps_as_pydicts="lossy") == {'a': 2}
```
Can we check that a warning is actually emitted? See [`pytest.warns`](https://docs.pytest.org/en/stable/how-to/capture-warnings.html#warns)

jonded94 replied:
Implemented a check for this warning

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -148,7 +149,26 @@ cdef class Scalar(_Weakrefable):
+        Parameters
+        ----------
+        maps_as_pydicts : str, optional, default `None`
+            Valid values are `None`, 'lossy', or 'strict'.
```
Like the `MonthDayNanoIntervalArray` comment, we don't need to explain `maps_as_pydicts` for non-nested Scalar subclasses. We can just mention it's ignored here.

jonded94 replied:
This is the base class, I would include an explanation here, as `List`, `Map`, `Struct`, ... are nested Scalars which inherit from this class.

But let me know what you think.

pitrou replied:
Oops, sorry. I misplaced this comment then :)

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -880,12 +1348,49 @@ cdef class MapScalar(ListScalar):
+        for key, value in self:
+            if key in result_dict:
+                if maps_as_pydicts == "strict":
+                    raise ValueError(
```
I would make this a `KeyError`. Also, the message should perhaps contain the duplicate key?

pitrou commented:
@github-actions crossbow submit -g python

pitrou reviewed (approved):
+1, will merge if CI is green.

Linchin commented:
Just fyi this might cause backward incompatibility issue because the user defined extension types are not expecting `maps_as_pydicts` as an argument for `to_py()`. We are seeing this in our prerelease tests: 
```
_________________________ test_json_arrow_record_batch _________________________

    def test_json_arrow_record_batch():
        data = [
            json.dumps(value, sort_keys=True, separators=(",", ":"))
            for value in JSON_DATA.values()
        ]
        arr = pa.array(data, type=db_dtypes.JSONArrowType())
        batch = pa.RecordBatch.from_arrays([arr], ["json_col"])
        sink = pa.BufferOutputStream()
    
        with pa.RecordBatchStreamWriter(sink, batch.schema) as writer:
            writer.write_batch(batch)
    
        buf = sink.getvalue()
    
        with pa.ipc.open_stream(buf) as reader:
            result = reader.read_all()
    
        json_col = result.column("json_col")
        assert isinstance(json_col.type, db_dtypes.JSONArrowType)
    
>       s = json_col.to_pylist()

tests/unit/test_json.py:225: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pyarrow/table.pxi:1380: in pyarrow.lib.ChunkedArray.to_pylist
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: JSONArrowScalar.as_py() got an unexpected keyword argument 'maps_as_pydicts'

pyarrow/array.pxi:1[67](https://github.com/googleapis/python-db-dtypes-pandas/actions/runs/13017353125/job/37736481135?pr=310#step:5:68)7: TypeError
- generated xml file: /home/runner/work/python-db-dtypes-pandas/python-db-dtypes-pandas/unit_prerelease_3.12_sponge_log.xml -
=========================== short test summary info ============================
FAILED tests/unit/test_json.py::test_json_arrow_to_pylist - TypeError: JSONArrowScalar.as_py() got an unexpected keyword argument 'maps_as_pydicts'
FAILED tests/unit/test_json.py::test_json_arrow_record_batch - TypeError: JSONArrowScalar.as_py() got an unexpected keyword argument 'maps_as_pydicts'
2 failed, 298 passed in 1.[81](https://github.com/googleapis/python-db-dtypes-pandas/actions/runs/13017353125/job/37736481135?pr=310#step:5:82)s
```
(Link: https://github.com/googleapis/python-db-dtypes-pandas/actions/runs/13017353125/job/37736481135?pr=310)

pitrou commented:
@Linchin @omatthew98 I think the way around this would be to take a `**kwargs` in your `as_py` method and then forward it to any nested `as_py` call (if any).

For example turn this:
```python
class JSONArrowScalar(pa.ExtensionScalar):
    def as_py(self):
        return JSONArray._deserialize_json(self.value.as_py() if self.value else None)
```

into this:
```python
class JSONArrowScalar(pa.ExtensionScalar):
    def as_py(self, **kwargs):
        return JSONArray._deserialize_json(self.value.as_py(**kwargs) if self.value else None)
```

pitrou commented:
I've updated the PR description, we should remember to call out this potential incompatibility in the release notes for the next version.
